packages/custom-functions-metadata/src/parseTree.ts

```diff
@@ -177,30 +198,129 @@ export function parseTree(sourceCode: string, sourceFileName: string): IParseTre
   const metadataFunctionNames: string[] = [];
-  const ids: string[] = [];
+  const metadataFunctionIds: string[] = [];
+  const metadataEnumIds: string[] = [];
 
   const sourceFile = ts.createSourceFile(sourceFileName, sourceCode, ts.ScriptTarget.Latest, true);
 
   buildEnums(sourceFile);
-  visit(sourceFile);
+  buildCustomEnums(sourceFile);
+  buildFunctions(sourceFile);
   const parseTreeResult: IParseTreeResult = {
     associate,
     extras,
     functions,
+    enums: customEnums,
   };
   return parseTreeResult;
 
+  // Build list for basic enums without tag "@customenum"
   function buildEnums(node: ts.Node) {
-    if (ts.isEnumDeclaration(node)) {
-      enumList.push(node.name.getText());
+    if (ts.isEnumDeclaration(node) && !isCustomEnum(node)) {
+      basicEnums.push(node.name.getText());
     }
     ts.forEachChild(node, buildEnums);
   }
+  
+  // Build list for custom enums tagged with "@customenum"
+  function buildCustomEnums(node: ts.Node) {
+    if (ts.isEnumDeclaration(node)) {
+      buildSingleCustomEnum(node);
+    }
+    ts.forEachChild(node, buildCustomEnums);
+  }
+
+  function buildSingleCustomEnum(node: ts.Node) {
+    if (!ts.isEnumDeclaration(node) || !node.parent || node.parent.kind !== ts.SyntaxKind.SourceFile) {
+      return;
+    }
+
+    const enumDeclaration = node as ts.EnumDeclaration;
+    if (!isCustomEnum(enumDeclaration) || enumDeclaration.members.length === 0) {
+      return;
+    }
+
+    const extra: IFunctionExtras = {
+      errors: [],
+      javascriptFunctionName: "",
+    };
+
+    const id = enumDeclaration.name.text;
+    validateId(id, getPosition(enumDeclaration), extra);
+    extras.push(extra);
+
+    if (checkForDuplicate(metadataEnumIds, id)) {
+      const errorString = `@${CUSTOM_ENUM} tag specifies a duplicate name: ${id}`;
+      extras.push({ errors: [logError(errorString, getPosition(enumDeclaration))], javascriptFunctionName: "" });
+    }
+
+    metadataEnumIds.push(id);
+
+    let isNumberEnum = true; // Default true for enums without initializer
+
+    // Extract JSDoc type from the enum declaration
+    let jsDocType: string | null = null;
+    const comment = getTagComment(enumDeclaration, CUSTOM_ENUM);
+    if (comment) {
+      const typeMatch = comment?.match(/\{\s*(string|number)\s*\}/);
+      if (!typeMatch) {
+        const errorString = `Unknown enum type defined after @${CUSTOM_ENUM} tag. Please use "{string}" or "{number}": ${id}`;
+        extras.push({ errors: [logError(errorString, getPosition(enumDeclaration))], javascriptFunctionName: "" });
+      }
+
+      jsDocType = typeMatch? typeMatch[1].trim() : null;
+    }
+
+    // Get the first member of the enum to determine if it is a number or string enum
+    const firstMember = enumDeclaration.members[0];
+    if (firstMember.initializer) {
+      const initializerText = firstMember.initializer.getText();
+      isNumberEnum = !isNaN(Number(initializerText));
+      if (jsDocType) {
```

Comment:
Do we want this to have an `else`?
Should handle if `jsDocType == null`, and `isNumberEnum == true`, in this case `jsDocType` should be set to `number` ?

Reply:
That is unnecessary because jsDocType won't be used anymore after this `if`. 
`isNumberEnum` is the flag in following logic and that is just why the type in JsDoc tag is optional.